### PR TITLE
Reworked git-meta documentation.

### PR DESCRIPTION
**Pull request by bpeabody**
This is the first step in changing the mission of git-meta to be a mechanism
for implementing a mono-repo.

I have trimmed down the `README.md` file to provide a basic introduction and
quick-start for git-meta.  I've split out an architectural whitepaper into
`doc/architecture.md` and added references in `README.md` to future documents
such as user and administration guides.

### CODE REVIEW

**Comment by bpeabody** [comment]
Note that there is more documentation to be written, my intention is to:
- replace existing documentation that is incompatible with our new mission
- provide architectural white paper describing our intentions and how we got there (before we forget!)
- restructure the README.md to have placeholders for other documentation (e.g., user guide) that needs to be written

Note also that I may replace the beautiful ASCII art at some point.

File: README.md [hunk 1]
```diff
@@ -29,179 +29,55 @@
...
-addressing shortcomings as we can, but Git-meta is not officially supported by
-Two Sigma at this time.
+## TL;DR
 
-Efficiently manage a set of repositories as if they were a single repository.
+Git-meta allows developers to work with extremely large codebases --
+performance degrades very slowly with increases in physical size, number of
+files, number of contributors, and depth of history.  Users clone only the
+subsets of the code that they need, yet can still make atomic commits across
+the entire codebase.  Development and collaboration are done using mostly
```
Comment (nkicillof) [inline README.md]: Another advantage could be having different ACLs for different repos

Comment (bpeabody) [inline README.md]: Yes, I had that advantage in there at some point; will put it back.

File: doc/architecture.md [hunk 2]
```diff
@@ -0,0 +1,967 @@
...
+
+Therefore, ref names in git-meta always refer to refs in the meta-repo.
+Branches and tags in sub-repos are ignored by git-meta -- by server-side checks
+and by the client-side `git-meta` plugin.  You may choose to create branches or
+tags in sub-repos (for example, to mirror significant branches and tags in a
+meta-repo), but they will not affect git-meta.  The git-meta plugin does not
+push branches or tags in sub-repos when, e.g. the `git meta push` command is
+used.
+
+## Syntetic-meta-refs
```
Comment (jhedwardyang) [inline doc/architecture.md]: Synthetic.

File: doc/architecture.md [hunk 3]
```diff
@@ -0,0 +1,967 @@
...
+# Performance
+
+At a minimum, users working in a mono-repo must download the meta-repo and all
+sub-repos containing code that they require to work.
+
+There is a commit in the meta-repo for every change made in the organization,
+so the number of commits in the history of the meta-repo may be very large.
+However, the information contained in each commit is relatively small,
+generally indicating only changes to submodule pointers.  Furthermore, the
+on-disk (checked out) rendering of the meta-repo is also small, being only a
```
Comment (jhedwardyang) [inline doc/architecture.md]: Perhaps mention shallow clones - this significantly speeds up our cloning since our cost is the history, not the working directory.

File: README.md [hunk 4]
```diff
@@ -29,179 +29,57 @@
...
-**NOTE Git-meta is BETA software**: Git-meta is open for collaboration, but
-currently in a very early phase of development.  We will be adding features and
-addressing shortcomings as we can, but Git-meta is not officially supported by
-Two Sigma at this time.
+## TL;DR
 
-Efficiently manage a set of repositories as if they were a single repository.
+Git-meta allows developers to work with extremely large codebases --
+performance only degrades very slowly when physical size, number of
+files, number of contributors increases, or the depth of history grows.
```
Comment (novalis) [inline README.md]: s/files,/files, or/

File: doc/architecture.md [hunk 5]
```diff
@@ -0,0 +1,967 @@
...
+- number of developers
+
+## What are the advantages of a mono-repo?
+
+The alternative to a mono-repo is for an organization to decompose its source
+into multiple repositories.  In comparison to a multi-repo strategy,
+a mono-repo provides the following advantages:
+
+- Atomic changes can be made across the organization's code.
+- The history of the of an organization's source is described in a mono-repo.
```
Comment (novalis) [inline doc/architecture.md]: s/of the of/of/

File: doc/architecture.md [hunk 6]
```diff
@@ -0,0 +1,967 @@
...
+- many are strongly focused on supporting a specific software platform (Android
+  Repo, gclient)
+- doesn't fully solve the scaling issue (Git subtrees)
+- prohibitively difficult to use (Git submodules)
+- lack scalable collaboration (e.g., pull request) strategies
+
+Git submodules come the closest: they do provide the technical ability to solve
+the problem, but are very difficult to use and lack some of the desired
+features.  With git-meta, we build on top of Git submodules to provide the
+desired functionality leveraging existing Git commands.
```
Comment (novalis) [inline doc/architecture.md]: s/leveraging/by leveraging/?  Or something -- this sentence is hard to parse.

**Comment by bpeabody** [comment]
I think I've addressed all concerns so far.

File: doc/architecture.md [hunk 7]
```diff
@@ -0,0 +1,967 @@
...
+| [m1]       | b [b2] |  | [a1]   |  | [b1]   |
+`---------------------,  `--------,  `--------,
+```
+
+We could have an automated task that would detect the creation of new
+repositories and auto-fork them, but when?  To allow for collaboration, we
+would most likely need to perform the auto-fork whenever a new repository is
+created, likely a very expensive operation for a potentially speculative
+operation.  The existence of these forks could be confusing, at best, if the
+when repositories are abandoned.  At the very least, we have created a new
```
Comment (novalis) [inline doc/architecture.md]: s/when/new/

File: doc/architecture.md [hunk 8]
```diff
@@ -0,0 +1,967 @@
...
+## Conclusions
+
+1. It is not generally possible to synchronize or validate updates to a ref
+   with the same name across many repositories.  Therefore, we treat ref names
+   in only meta-repos as significant; git-meta does not push branches or tags
+   in sub-repos.
+1. We use symbolic-meta-refs as push-targets in sub-repos; as the contents of a
+   symbolic-meta-ref are immutable (a given symbolic-meta-ref can point to only
+   one commit), we are always guaranteed to be able to update them when needed.
+1. Forking sub-repos is potentially expensive, an probably impractical on the
```
Comment (novalis) [inline doc/architecture.md]: s/an /and /
